File: solutions/backtracking/040-combination-sum-ii.py
"""
# 040. Combination Sum Ii
**Medium**

Given a problem that demonstrates key concepts in Backtracking.

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of backtracking concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply backtracking methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages backtracking principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### TIME COMPLEXITY: O(n)
### SPACE COMPLEXITY: O(1)

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>

<details>
<summary><b>💡 APPROACH</b></summary>

The approach uses backtracking techniques to solve this problem efficiently.

### Algorithm Steps:
1. Initialize necessary variables
2. Process input using backtracking method
3. Return the computed result

</details>
"""

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        """
        Find all unique combinations of numbers that sum to target.
        
        Args:
            candidates: List of candidate numbers
            target: Target sum to achieve
            
        Returns:
            List of all unique combinations that sum to target
        """
        # Sort candidates to handle duplicates and enable early termination
        candidates.sort()
        result = []
        
        def backtrack(curr: List[int], pos: int, remain: int) -> None:
            """
            Recursive backtracking helper function.
            
            Args:
                curr: Current combination being built
                pos: Current position in candidates array
                remain: Remaining sum to achieve
            """
            if remain == 0:
                # Found a valid combination
                result.append(curr[:])
                return
            
            for i in range(pos, len(candidates)):
                # Skip duplicates to avoid duplicate combinations
                if i > pos and candidates[i] == candidates[i-1]:
                    continue
                    
                # Early termination if current number is too large
                if candidates[i] > remain:
                    break
                    
                # Include current number and recurse
                curr.append(candidates[i])
                backtrack(curr, i + 1, remain - candidates[i])
                curr.pop()  # Backtrack by removing the last number
        
        backtrack([], 0, target)
        return result

def test_solution():
    """
    Test cases for 040. Combination Sum Ii.
    """
    solution = Solution()

    # Test case 1: Basic functionality
    # result = solution.solve([test_input])
    # expected = [expected_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Edge case
    # result = solution.solve([edge_case_input])
    # expected = [edge_case_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 040. Combination Sum Ii")

File: solutions/backtracking/alternatives/040-combination-sum-ii.js
/**
 * 40. Combination Sum Ii
 * Medium
 *
 * Combination Sum II - Find all unique combinations of candidates where chosen numbers sum to target @param {number[]} candidates - Array of candidate numbers @param {number} target - Target sum @return {number[][]} - Array of all unique combinations that sum to target
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * The key insight for solving Combination Sum Ii is to understand the core problem pattern
 * and apply the most efficient algorithmic approach.
 *
 * ### APPROACH:
 * 1. Analyze the problem requirements
 * 2. Choose the optimal data structure
 * 3. Implement the solution step by step
 * 4. Handle edge cases appropriately
 *
 * ### WHY THIS WORKS:
 * This approach works because it leverages the fundamental properties of the problem
 * to achieve an efficient solution.
 *
 * ### EXAMPLE WALKTHROUGH:
 * For a typical input, the algorithm processes the data systematically
 * to produce the expected output.
 *
 * </details>
 */

/**
 * Combination Sum II - Find all unique combinations of candidates where chosen numbers sum to target
 * 
 * @param {number[]} candidates - Array of candidate numbers
 * @param {number} target - Target sum
 * @return {number[][]} - Array of all unique combinations that sum to target
 */
function combinationSum2(candidates, target) {
    // Sort the candidates array to handle duplicates and enable early termination
    candidates.sort((a, b) => a - b);
    
    // Result array to store all valid combinations
    const result = [];
    
    /**
     * Recursive backtracking function to find combinations
     * @param {number} remain - Remaining sum to reach target
     * @param {number[]} current - Current combination being built
     * @param {number} start - Starting index for considering candidates
     */
    function backtrack(remain, current, start) {
        // Base case: if remaining sum is 0, we found a valid combination
        if (remain === 0) {
            result.push([...current]);
            return;
        }
        
        // Iterate through candidates starting from 'start' index
        for (let i = start; i < candidates.length; i++) {
            // Skip duplicates to avoid duplicate combinations
            if (i > start && candidates[i] === candidates[i - 1]) continue;
            
            // If current candidate is greater than remaining sum, break
            // (since array is sorted, all following numbers will be too large)
            if (candidates[i] > remain) break;
            
            // Include current candidate in combination
            current.push(candidates[i]);
            
            // Recursively find combinations with remaining sum
            // Note: i + 1 ensures we don't reuse the same element
            backtrack(remain - candidates[i], current, i + 1);
            
            // Backtrack: remove current candidate to try next possibility
            current.pop();
        }
    }
    
    // Start backtracking with initial values
    backtrack(target, [], 0);
    
    return result;
}

/**
 * Example usage and test cases
 */
function runTests() {
    // Test Case 1
    console.log("Test 1:");
    console.log(combinationSum2([10,1,2,7,6,1,5], 8));
    // Expected: [[1,1,6], [1,2,5], [1,7], [2,6]]
    
    // Test Case 2
    console.log("\nTest 2:");
    console.log(combinationSum2([2,5,2,1,2], 5));
    // Expected: [[1,2,2], [5]]
    
    // Test Case 3: Empty array
    console.log("\nTest 3:");
    console.log(combinationSum2([], 1));
    // Expected: []
    
    // Test Case 4: No possible combinations
    console.log("\nTest 4:");
    console.log(combinationSum2([1], 2));
    // Expected: []
}

// Export the function for potential module usage
module.exports = {
    combinationSum2
};

// Run tests if file is executed directly
if (require.main === module) {
    runTests();
}